### PR TITLE
BFS

### PR DESCRIPTION
**Pull request by spacejam**

### CODE REVIEW

File: src/query.rs [hunk 1]
```diff
@@ -120,17 +42,65 @@ impl QueryEngine {
...
+    async fn traverse(&mut self, traverse: Traverse) -> io::Result<QueryResult> {
+        match traverse.kind {
+            TraversalKind::Bfs => self.bfs(traverse.path_filter).await,
+            TraversalKind::Dfs => self.dfs(traverse.path_filter).await,
         }
+    }
+
+    async fn bfs(&mut self, path_filter: PathFilter) -> io::Result<QueryResult> {
+        let mut tx = self.cluster_client.begin_transaction().await?;
+        let l0 = if let Ok(l0) = tx.scan(path_filter.start).await {
```
Comment (gitbuda) [inline src/query.rs:54]: `l0` means "level 0"?

**Comment by spacejam** [comment]
just as a note for future comparison, before any optimizations happen, doing BFS on the FB dataset from the first ID to the last currently takes 19 seconds with the most naive implementation possible, with 
```
SimulatorStats {
  simulator_ticks: 395527422,
  messages_sent: 5908,
  requests_sent: 2954 
}
```

**Comment by spacejam** [comment]
after adding edge and node lookup specializations, BFS now takes 216ms. So, ~100x faster

File: src/common.rs [hunk 2]
```diff
@@ -138,3 +136,118 @@ impl Address {
         (self.ip_addr, self.port)
     }
 }
+
+/// This type is only for returning data to users, not internal representation
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Vertex {
+    pub label: String,
```
[the remark this answers is not in the document]
Comment (spacejam) [reply src/common.rs:85]: we can expand that in the future

File: src/query.rs (hunk 1 above)
Comment (spacejam) [reply src/query.rs]: yes

**Pull request merged**